heat/core/linalg/basics.py

```diff
@@ -44,6 +45,118 @@
+
+    Parameters
+    ----------
+    x : DNDarray
+        Symmetric or Hermitian positive-definite matrix. Shape = (...,M,M)
+    upper : bool, optional
+        Whether a lower or upper triangular matrix is returned.
+
+    Note
+    ----
+    Distributed matrices with complex valued entries are currently not supported.
+
+    Examples
+    --------
+    >>> A = ht.array([[1,-2j],[2j,5]])
+    >>> A
+    DNDarray([[ (1+0j), (-0-2j)],
+              [     2j,  (5+0j)]], dtype=ht.complex64, device=cpu:0, split=None)
+    >>> L = ht.linalg.cholesky(A)
+    >>> L
+    DNDarray([[(1+0j),     0j],
+              [    2j, (1+0j)]], dtype=ht.complex64, device=cpu:0, split=None)
+    >>> L @ L.T.conj()
+    DNDarray([[(1+0j),    -2j],
+              [    2j, (5+0j)]], dtype=ht.complex64, device=cpu:0, split=None)
+    """
+    if x.ndim < 2:
+        raise RuntimeError("DNDarray must be at least two-dimensional.")
+
+    m, n = x.shape[-2:]
+
+    if m != n:
+        raise RuntimeError("Last two dimensions of the DNDarray must be square.")
+
+    if types.heat_type_is_exact(x.dtype):
+        raise NotImplementedError("Not implemented for {}".format(x.dtype))
+
+    # matrix  single process
+    if not x.is_distributed() or x.split < x.ndim - 2:
+        try:
+            data = torch.linalg.cholesky(x.larray, upper=upper)
+        except TypeError:  # upper parameter first introduced in PyTorch 1.10.
+            data = torch.linalg.cholesky(x.larray)
+            if upper:
+                data = data.T.conj()
+        except AttributeError:  # PyTorch 1.7. has nearly no linalg module
+            data = torch.cholesky(x.larray)
+            if upper:
+                data = data.T.conj()
+
+        return DNDarray(
+            data, x.shape, types.heat_type_of(data), x.split, x.device, x.comm, x.balanced
+        )
+
+    # NOTE There is some issues with complex values. conj() is omitted in the computation.
+    if types.heat_type_is_complexfloating(x.dtype):
+        raise NotImplementedError(
+            "Not implemented for distributed matrices of type {}".format(x.dtype)
+        )
+
+    a = x.copy()
+
+    for i in range(n - 1):
+        a[..., i, i] = exponential.sqrt(a[..., i, i])
+
+        if a.split == a.ndim - 2:
+            if upper:
+                a[..., i, i + 1 :] /= manipulations.expand_dims(a[..., i, i], -1)
+                b = a[..., i, i + 1 :]
+                L = triu(
+                    manipulations.expand_dims(manipulations.resplit(b, -1), -1)
+                    * manipulations.expand_dims(b, -2)
+                )
+            else:
+                a[..., i + 1 :, i].larray /= manipulations.expand_dims(a[..., i, i], -1).larray
+                b = a[..., i + 1 :, i]
+                b.balance_()
+                L = tril(
+                    manipulations.expand_dims(b, -1)
+                    * manipulations.expand_dims(manipulations.resplit(b, None), -2)
+                )
```

Comment:
I find it problematic that `b` needs to be gathered. This is basically an outer product, isn't it? Distributed `outer` is implemented, and we should be able to expand that functionality to a stack of vectors.

(Also it shouldn't be necessary to balance just before resplitting, if it is let's fix it)